File: src/ios/core/Utilities/Sequences/Queue.swift
//
//  Queue.swift
//  WTVJavaScriptCore
//
//  Created by Jing KE on 17/06/16.
//  Copyright © 2016 WizTiVi. All rights reserved.
//

import Foundation

struct Queue<Element> {
    var items: Array<Element>
    var isEmpty: Bool {
        return self.items.isEmpty
    }
    var frontItem: Element? {
        return self.isEmpty ? nil : self.items[0]
    }
    init(items: Array<Element> = Array<Element>()) {
        self.items = items
    }
    mutating func enqueue(item: Element) {
        self.items.append(item)
    }
    mutating func dequeue() -> Element? {
        return self.isEmpty ? nil : self.items.removeAtIndex(0)
    }
    static func create() -> Queue<Element> {
        return Queue<Element>()
    }
}

File: www/core/Utilities/Sequences/Queue.js
class Queue {

    constructor() {
        this.items = [];
    }

    get isEmpty() {
        return this.items.length === 0;
    }

    get frontItem() {
        return this.isEmpty ? (void 0) : this.items[0];
    }

    enqueue() {
        for(var argument of arguments) {
            this.items.push(argument);
        }
    }

    dequeue() {
        return this.isEmpty ? (void 0) : this.items.shift();
    }

}